src/storage/src/upsert.rs

```diff
@@ -96,23 +91,29 @@ impl UpsertKey {
         Self::from_iter(key.map(|r| r.iter()))
     }
 
-    pub fn from_value(value: Result<&Row, &UpsertError>, mut key_indices: &[usize]) -> Self {
-        Self::from_iter(value.map(|value| {
-            value.iter().enumerate().flat_map(move |(idx, datum)| {
-                let key_idx = key_indices.get(0)?;
-                if idx == *key_idx {
-                    key_indices = &key_indices[1..];
-                    Some(datum)
-                } else {
-                    None
-                }
-            })
-        }))
+    pub fn from_value(value: Result<&Row, &UpsertError>, key_indices: &[usize]) -> Self {
+        thread_local! {
+            /// A thread-local datum cache used to calculate hashes
+            static VALUE_DATUMS: RefCell<DatumVec> = RefCell::new(DatumVec::new());
+        }
+        VALUE_DATUMS.with(|value_datums| {
```

Comment:
You could change `from_value` to accept a `&mut DatumVec` instead of using a thread-local. There seems to be a single callsite, so that change wouldn't be too big, and you'll probably get significant performance benefits. (thread-local variables need to do an initialization check on every get call.)

Reply:
Or, change it signature to `value: Result<DatumVecBorrow, ...>` and decode outside.

Reply:
> you'll probably get significant performance benefits

Hm, why is this the case? I'm reluctant to make a more complicated API without having actually measured the impact. Oftentimes what we think will have a negative performance impact turns out to be immaterial.

 In this specific case I think it would be difficult to even observe the performance impact of this check compared to the rest of the work this function is doing since it compiles down to a non-atomic check of a register and the branch will almost always be predicted correctly. This is the assembly I got from a [quick godbolt test ](https://godbolt.org/z/fx9c3jf1x)

```asm
        call    __tls_get_addr@PLT
        mov     rbx, rax
        mov     rax, qword ptr [rax + example::FOO::{{constant}}::{{closure}}::VAL::ha9d33ec14702ed6d@DTPOFF]
        cmp     rax, 1
        jne     .LBB2_1
```

Reply:
Cool, that's fine with me!